Improve helm dependency update performance

JvD-Ericsson opened the pull request:
What this PR does / why we need it:
This PR was created to improve performance of the dependency update command by skipping unnecessary downloading and loading of index files that have already been downloaded and loaded

I believe this would close refs #9865

Signed-off-by: Jeff van Dam <jeff.van.dam@est.tech>

JvD-Ericsson committed:
Improve helm dependency update performance

What this PR does / why we need it:
This PR was created to improve performance of the dependency update command by
skipping unnecessary downloading and loading of index files that have already
been downloaded and loaded

I believe this would close refs #9865

Signed-off-by: Jeff van Dam <jeff.van.dam@est.tech>

JvD-Ericsson committed:
Improve helm dependency update performance

What this PR does / why we need it:
This PR was created to improve performance of the dependency update command by
skipping unnecessary downloading and loading of index files that have already
been downloaded and loaded

I believe this would close refs #9865

Signed-off-by: Jeff van Dam <jeff.van.dam@est.tech>

joejulian commented:
Changed the milestone because this isn't a bug fix or security issue.

SuperSandro2000 commented:
Could I motivate people to review this? It would be much appreciated.

MichaelMorrisEst commented on pkg/downloader/manager.go:
```diff
@@ -735,7 +745,19 @@ func (m *Manager) findChartURL(name, version, repoURL string, repos map[string]*
+			baseChartUrls[repoURL] = strings.Join(slice[:len(slice)-1], "")
+		}
+	} else {
+		url = fmt.Sprintf("%s%s-%s.tgz", baseURL, name, version)
```
This makes assumptions about the structure of the chart URL, are we sure the assumptions are correct?

An alternative suggestion:
The chart URL is knowable earlier in the flow, in Resolve (resolver.go) at the point where we set the version (inside loop on lines 197-200), the url is available from ver.URLs[0]. We could store (perhaps in a map also returned from the method) the chart URLs which can be used during the download rather than using additional logic to again find the url

JvD-Ericsson replied:
Cool, yeah makes more sense to do it that way. Have that updated now

JvD-Ericsson committed:
Merge branch 'fixDepUpPerformance' of https://github.com/Nordix/helm into fixDepUpPerformance

joejulian commented on pkg/downloader/chart_downloader.go:
```diff
@@ -184,11 +183,11 @@ func (c *ChartDownloader) getOciURI(ref, version string, u *url.URL) (*url.URL,
-//		* If version is empty, this will return the URL for the latest version
-//		* If no version can be found, an error is returned
+//   - For fully qualified URLs, the version will be ignored (since URLs aren't versioned)
+//   - For a chart reference
+//   - If version is non-empty, this will return the URL for that version
+//   - If version is empty, this will return the URL for the latest version
+//   - If no version can be found, an error is returned
```
I think vscode reformatted this

JvD-Ericsson replied:
Yeah it annoyingly does it on save. Have it back as the original now

SuperSandro2000 replied:
This is probably due to go 1.20 using markdown for comments. A resave of the files probably triggers a go fmt which then generates this.

joejulian replied:
I have the same problem with this file and vscode. :smirk:

JvD-Ericsson committed:
Update chart_downloader.go

Signed-off-by: Jeff van Dam <66410239+JvD-Ericsson@users.noreply.github.com>

JvD-Ericsson committed:
Update chart_downloader.go

Signed-off-by: Jeff van Dam <66410239+JvD-Ericsson@users.noreply.github.com>

JvD-Ericsson commented:
Hey @joejulian any plan to merge this anytime soon? Happy to make any changes if necessary

JvD-Ericsson commented:
Sorry to bother again @joejulian but is there any chance of this being merged anytime soon?

joejulian commented on pkg/downloader/manager.go:
```diff
@@ -529,6 +530,14 @@ func (m *Manager) ensureMissingRepos(repoNames map[string]string, deps []*chart.
 
 		repoNames[dd.Name] = rn
 
+		// If repository is already present don't add to array. This will skip
+		// unnecessary index file downloading improving performance.
+		for _, item := range ru {
+			if item.URL == dd.Repository {
+				continue Outer
+			}
+		}
+
```
Will this change the behavior? I would imagine if I had a workflow that expected `helm dependency update` to grab the latest index.yaml every time, this would break that, would it not?

```
dependencies:
  - name: console
    condition: console.enabled
    version: ">=0.5 <1.0"
    repository: https://charts.redpanda.com
```

This change would keep that version constraint from updating the subchart if a new `console` chart was released, right?

MichaelMorrisEst replied:
Hi @joejulian 
This if statement only prevents the same repo being added multiple times to the array of repos to be updated as part of the dependency update, and therefore prevents the same repo being downloaded multiple times as part of the same dependency update. I believe the situation you mention will be handled as expected with the latest index.yaml being downloaded (but only once, irrespective of how many times it is referenced in the dependencies)

rolfberkenbosch commented:
Thank you @JvD-Ericsson for your work. We have currently a lot of performance problemen with ArgoCD in combination with Helm. We have investigated the performance problems and we found out that this is due the helm dependencies. We have tested this pr and it will improve the performance a lot! @joejulian can you merge this pr ?

joejulian reviewed (approved):
lgtm. I've tried some manual testing and everything seems to work. I don't have any examples that take a long time, but at least everything I use regularly seems to keep working.

mattfarina commented on pkg/downloader/chart_downloader.go:
```diff
@@ -378,19 +377,13 @@ func (c *ChartDownloader) scanReposForURL(u string, rf *repo.File) (*repo.Entry,
-					}
-				}
-			}
+		file := string(yamlFile[:])
+		if strings.Contains(file, u) {
+			return rc, nil
```
I'm not sure about this change. The `string.Contains` is run over a file with annotations and descriptions. The `u` could be in one of those fields which would return an improper result. While this is unlikely, the code that calls `scanReposForURL` grabs the credentials if there are any to pass on. I wonder if there is an edge case that could leak credentials with misuse of this change.

JvD-Ericsson replied:
I have that reverted now, so hopefully it can be merged now! If anything else needs changing let me know

mattfarina reviewed (changes requested):
This mostly makes sense from a performance perspective.

The one issue I found is a way to possibly send credentials to the wrong URL by including the URL the credentials are associated with in an annotation or description in a malicious index.yaml file. I think that part needs to be walked back for the sake of security.

JvD-Ericsson committed:
Merge branch 'fixDepUpPerformance' of https://github.com/Nordix/helm into fixDepUpPerformance

lapastillaroja commented:
It would be great to get this PR merged!

mattfarina commented on pkg/downloader/manager.go:
```diff
@@ -735,7 +744,13 @@ func (m *Manager) findChartURL(name, version, repoURL string, repos map[string]*
 	}
-	url, err = repo.FindChartInRepoURL(repoURL, name, version, certFile, keyFile, caFile, m.Getters)
+
+	if _, ok := urls[name]; ok {
+		url = urls[name]
+	} else {
+		url, err = repo.FindChartInRepoURL(repoURL, name, version, certFile, keyFile, caFile, m.Getters)
+	}
```
`name` is not globally unique. There could be a chart with the same name in two different repos. And, in a chart there could be multiple versions used. The key lookup needs to deal with this. While somewhat rare, I've encountered what I just described. The cache key needs to account for the repo, name, and version. Just like the `FindChartInRepoURL` function does. This starts in the `resolver.Resolve` function. I point it out here because this is where the impact is felt.

JvD-Ericsson replied:
@mattfarina hopefully the key including the repo, name, version should suffice

JvD-Ericsson replied:
If anymore changes are needed just let me know. Hopefully this can get merged before next release

mattfarina reviewed (changes requested):
I'm sorry I didn't catch this with my last review. Caching is hard and I just realized the impact of cache key possibilities.

JvD-Ericsson committed:
Made urls key more specific

Signed-off-by: Jeff van Dam <jeff.van.dam@est.tech>

robbertuittenbroek commented:
@mattfarina subtle nudge to please review the changes made by @JvD-Ericsson. The milestone for which this issue is scheduled is due en like some others we are eager for this fix. Thanks!

rolfberkenbosch commented:
@joejulian or @mattfarina I saw that 3.13.0 is just released without this pr. Will this one be merged in het 3.14.0 release ?

scottrigby reviewed (approved):
Hey everyone, I'm a little late to the party. Happy holidays.

@JvD-Ericsson really nice work 👏 

Yes 183f01b32c9bcf520ca17128309347753bae2d6f addresses what @mattfarina requested in https://github.com/helm/helm/pull/11726#pullrequestreview-1605482574. Glad he pointed this out.

Tested and all lgtm

robbertuittenbroek commented:
Hi! Since @scottrigby reviewed and approved, I hope @joejulian and @mattfarina can give their blessing as well. This ticket is almost 1 year underway, feels like a good moment to wrap it up!

JvD-Ericsson commented:
@scottrigby, @mattfarina, @joejulian can this be merged? I have two approvals so seems like it should be good to go. I am happy to make any changes if necessary

MichaelMorrisEst commented:
> What is left to be done here? Are there still some coding tasks left?

Its waiting for review from maintainers. Already has one maintainer approval, but two needed before can be merged

sabre1041 reviewed (changes requested):
Logically, this looks good, however, I am receiving the following errors when running `make test-style`

```
make test-style
GO111MODULE=on golangci-lint run
WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
internal/test/ensure/ensure.go:60: File is not `gofmt`-ed with `-s` (gofmt)
//      tempdir := TempFile(t, "foo", []byte("bar"))
//      defer os.RemoveAll(tempdir)
//      filename := filepath.Join(tempdir, "foo")
pkg/lint/support/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
/*Package support contains tools for linting charts.
pkg/gates/doc.go:16: File is not `gofmt`-ed with `-s` (gofmt)
/*Package gates provides a general tool for working with experimental feature gates.
pkg/chartutil/dependencies.go:48:13: superfluous-else: if block ends with a break statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary) (revive)
                                        } else {
                                                log.Printf("Warning: Condition path '%s' for chart %s returned non-bool value", c, r.Name)
                                        }
pkg/releaseutil/kind_sorter.go:133:17: unused-parameter: parameter 'a' seems to be unused, consider removing or renaming it as _ (revive)
func lessByKind(a interface{}, b interface{}, kindA string, kindB string, o KindSortOrder) bool {
                ^
pkg/lint/rules/template.go:48:89: unused-parameter: parameter 'strict' seems to be unused, consider removing or renaming it as _ (revive)
func Templates(linter *support.Linter, values map[string]interface{}, namespace string, strict bool) {
                                                                                        ^
pkg/storage/driver/mock_test.go:252:38: unused-parameter: parameter 'releases' seems to be unused, consider removing or renaming it as _ (revive)
func newTestFixtureSQL(t *testing.T, releases ...*rspb.Release) (*SQL, sqlmock.Sqlmock) {
                                     ^
pkg/pusher/pusher.go:92:10: unused-parameter: parameter 'settings' seems to be unused, consider removing or renaming it as _ (revive)
func All(settings *cli.EnvSettings) Providers {
         ^
pkg/storage/storage_test.go:296:39: unused-parameter: parameter 'key' seems to be unused, consider removing or renaming it as _ (revive)
func (d *MaxHistoryMockDriver) Delete(key string) (*rspb.Release, error) {
                                      ^
pkg/plugin/installer/http_installer_test.go:47:30: unused-parameter: parameter 'href' seems to be unused, consider removing or renaming it as _ (revive)
func (t *TestHTTPGetter) Get(href string, _ ...getter.Option) (*bytes.Buffer, error) {
                             ^
cmd/helm/search/search_test.go:104:20: unused-parameter: parameter 't' seems to be unused, consider removing or renaming it as _ (revive)
func loadTestIndex(t *testing.T, all bool) *Index {
                   ^
pkg/repo/index_test.go:436:2: redefines-builtin-id: redefinition of the built-in function real (revive)
        var expected, real []string
        ^
pkg/repo/index_test.go:444:3: redefines-builtin-id: redefinition of the built-in function real (revive)
                real = append(real, scanner.Text())
                ^
pkg/repo/chartrepo_test.go:120:41: unused-parameter: parameter 'options' seems to be unused, consider removing or renaming it as _ (revive)
func (g *CustomGetter) Get(href string, options ...getter.Option) (*bytes.Buffer, error) {
                                        ^
pkg/kube/fake/printer.go:51:63: unused-parameter: parameter 'reader' seems to be unused, consider removing or renaming it as _ (revive)
func (p *PrintingKubeClient) Get(resources kube.ResourceList, reader io.Reader) (map[string][]runtime.Object, error) {
                                                              ^
pkg/kube/client.go:152:46: unused-parameter: parameter 'reader' seems to be unused, consider removing or renaming it as _ (revive)
func (c *Client) Get(resources ResourceList, reader io.Reader) (map[string][]runtime.Object, error) {
                                             ^
pkg/action/package.go:58:36: unused-parameter: parameter 'vals' seems to be unused, consider removing or renaming it as _ (revive)
func (p *Package) Run(path string, vals map[string]interface{}) (string, error) {
                                   ^
pkg/action/registry_logout.go:36:30: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
func (a *RegistryLogout) Run(out io.Writer, hostname string) error {
                             ^
pkg/action/registry_login.go:38:29: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
func (a *RegistryLogin) Run(out io.Writer, hostname string, username string, password string, insecure bool) error {
                            ^
cmd/helm/plugin_list.go:78:22: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
func compListPlugins(toComplete string, ignoredPluginNames []string) []string {
                     ^
cmd/helm/docs.go:80:27: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
func (o *docsOptions) run(out io.Writer) error {
                          ^
cmd/helm/history.go:187:24: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
func compListRevisions(toComplete string, cfg *action.Configuration, releaseName string) ([]string, cobra.ShellCompDirective) {
                       ^
cmd/helm/flags.go:196:39: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
func compVersionFlag(chartRef string, toComplete string) ([]string, cobra.ShellCompDirective) {
                                      ^
cmd/helm/repo_list.go:126:20: unused-parameter: parameter 'prefix' seems to be unused, consider removing or renaming it as _ (revive)
func compListRepos(prefix string, ignoredRepoNames []string) []string {
                   ^
cmd/helm/completion.go:213:20: unused-parameter: parameter 'cmd' seems to be unused, consider removing or renaming it as _ (revive)
func noCompletions(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
                   ^
pkg/downloader/manager.go:726:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
                                return
                                ^
pkg/downloader/manager.go:731:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
                                return
                                ^
pkg/downloader/manager.go:735:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
                                return
                                ^
make: *** [test-style] Error 1```

JvD-Ericsson committed:
Merge branch 'helm:main' into fixDepUpPerformance

JvD-Ericsson committed:
Update manager_test.go

Signed-off-by: Jeff van Dam <jeff.van.dam@est.tech>

JvD-Ericsson commented:
> Logically, this looks good, however, I am receiving the following errors when running `make test-style`
> 
> ```
> make test-style
> GO111MODULE=on golangci-lint run
> WARN [runner] The linter 'deadcode' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
> WARN [runner] The linter 'varcheck' is deprecated (since v1.49.0) due to: The owner seems to have abandoned the linter. Replaced by unused. 
> internal/test/ensure/ensure.go:60: File is not `gofmt`-ed with `-s` (gofmt)
> //      tempdir := TempFile(t, "foo", []byte("bar"))
> //      defer os.RemoveAll(tempdir)
> //      filename := filepath.Join(tempdir, "foo")
> pkg/lint/support/doc.go:17: File is not `gofmt`-ed with `-s` (gofmt)
> /*Package support contains tools for linting charts.
> pkg/gates/doc.go:16: File is not `gofmt`-ed with `-s` (gofmt)
> /*Package gates provides a general tool for working with experimental feature gates.
> pkg/chartutil/dependencies.go:48:13: superfluous-else: if block ends with a break statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary) (revive)
>                                         } else {
>                                                 log.Printf("Warning: Condition path '%s' for chart %s returned non-bool value", c, r.Name)
>                                         }
> pkg/releaseutil/kind_sorter.go:133:17: unused-parameter: parameter 'a' seems to be unused, consider removing or renaming it as _ (revive)
> func lessByKind(a interface{}, b interface{}, kindA string, kindB string, o KindSortOrder) bool {
>                 ^
> pkg/lint/rules/template.go:48:89: unused-parameter: parameter 'strict' seems to be unused, consider removing or renaming it as _ (revive)
> func Templates(linter *support.Linter, values map[string]interface{}, namespace string, strict bool) {
>                                                                                         ^
> pkg/storage/driver/mock_test.go:252:38: unused-parameter: parameter 'releases' seems to be unused, consider removing or renaming it as _ (revive)
> func newTestFixtureSQL(t *testing.T, releases ...*rspb.Release) (*SQL, sqlmock.Sqlmock) {
>                                      ^
> pkg/pusher/pusher.go:92:10: unused-parameter: parameter 'settings' seems to be unused, consider removing or renaming it as _ (revive)
> func All(settings *cli.EnvSettings) Providers {
>          ^
> pkg/storage/storage_test.go:296:39: unused-parameter: parameter 'key' seems to be unused, consider removing or renaming it as _ (revive)
> func (d *MaxHistoryMockDriver) Delete(key string) (*rspb.Release, error) {
>                                       ^
> pkg/plugin/installer/http_installer_test.go:47:30: unused-parameter: parameter 'href' seems to be unused, consider removing or renaming it as _ (revive)
> func (t *TestHTTPGetter) Get(href string, _ ...getter.Option) (*bytes.Buffer, error) {
>                              ^
> cmd/helm/search/search_test.go:104:20: unused-parameter: parameter 't' seems to be unused, consider removing or renaming it as _ (revive)
> func loadTestIndex(t *testing.T, all bool) *Index {
>                    ^
> pkg/repo/index_test.go:436:2: redefines-builtin-id: redefinition of the built-in function real (revive)
>         var expected, real []string
>         ^
> pkg/repo/index_test.go:444:3: redefines-builtin-id: redefinition of the built-in function real (revive)
>                 real = append(real, scanner.Text())
>                 ^
> pkg/repo/chartrepo_test.go:120:41: unused-parameter: parameter 'options' seems to be unused, consider removing or renaming it as _ (revive)
> func (g *CustomGetter) Get(href string, options ...getter.Option) (*bytes.Buffer, error) {
>                                         ^
> pkg/kube/fake/printer.go:51:63: unused-parameter: parameter 'reader' seems to be unused, consider removing or renaming it as _ (revive)
> func (p *PrintingKubeClient) Get(resources kube.ResourceList, reader io.Reader) (map[string][]runtime.Object, error) {
>                                                               ^
> pkg/kube/client.go:152:46: unused-parameter: parameter 'reader' seems to be unused, consider removing or renaming it as _ (revive)
> func (c *Client) Get(resources ResourceList, reader io.Reader) (map[string][]runtime.Object, error) {
>                                              ^
> pkg/action/package.go:58:36: unused-parameter: parameter 'vals' seems to be unused, consider removing or renaming it as _ (revive)
> func (p *Package) Run(path string, vals map[string]interface{}) (string, error) {
>                                    ^
> pkg/action/registry_logout.go:36:30: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
> func (a *RegistryLogout) Run(out io.Writer, hostname string) error {
>                              ^
> pkg/action/registry_login.go:38:29: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
> func (a *RegistryLogin) Run(out io.Writer, hostname string, username string, password string, insecure bool) error {
>                             ^
> cmd/helm/plugin_list.go:78:22: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
> func compListPlugins(toComplete string, ignoredPluginNames []string) []string {
>                      ^
> cmd/helm/docs.go:80:27: unused-parameter: parameter 'out' seems to be unused, consider removing or renaming it as _ (revive)
> func (o *docsOptions) run(out io.Writer) error {
>                           ^
> cmd/helm/history.go:187:24: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
> func compListRevisions(toComplete string, cfg *action.Configuration, releaseName string) ([]string, cobra.ShellCompDirective) {
>                        ^
> cmd/helm/flags.go:196:39: unused-parameter: parameter 'toComplete' seems to be unused, consider removing or renaming it as _ (revive)
> func compVersionFlag(chartRef string, toComplete string) ([]string, cobra.ShellCompDirective) {
>                                       ^
> cmd/helm/repo_list.go:126:20: unused-parameter: parameter 'prefix' seems to be unused, consider removing or renaming it as _ (revive)
> func compListRepos(prefix string, ignoredRepoNames []string) []string {
>                    ^
> cmd/helm/completion.go:213:20: unused-parameter: parameter 'cmd' seems to be unused, consider removing or renaming it as _ (revive)
> func noCompletions(cmd *cobra.Command, args []string, toComplete string) ([]string, cobra.ShellCompDirective) {
>                    ^
> pkg/downloader/manager.go:726:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
>                                 return
>                                 ^
> pkg/downloader/manager.go:731:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
>                                 return
>                                 ^
> pkg/downloader/manager.go:735:5: naked return in func `findChartURL` with 45 lines of code (nakedret)
>                                 return
>                                 ^
> make: *** [test-style] Error 1```
> ```

Should be fixed now

sabre1041 reviewed (approved):
LGTM

MichaelMorrisEst commented:
Thanks a million for reviewing @sabre1041, @scottrigby , can this now be merged as it has 2 maintainer approvals

scottrigby commented:
3.15.0 was already released, so I just moved this PR to the 3.16.0 milestone.

joejulian commented on pkg/downloader/manager.go:
```diff
@@ -191,13 +191,13 @@ func (m *Manager) Update() error {
+	lock, urls, err := m.resolve(req, repoNames)
 	if err != nil {
 		return err
 	}
```
Do we need to check that `urls` is empty?

MichaelMorrisEst replied:
`urls` is ultimately read on line 756, if it is empty the key will not be found and the flow will proceed to the else statement and proceed as per existing implementation without any issues, so I am not seeing a reason to check for empty

MichaelMorrisEst committed:
Merge branch 'helm:main' into fixDepUpPerformance

Signed-off-by: MichaelMorris <michael.morris@est.tech>

mattfarina commented:
@joejulian @scottrigby @sabre1041 can you re-review since new changes were pushed?

MichaelMorrisEst commented:
Thanks @joejulian for reviewing/merging and also everyone else who reviewed.
Thanks @JvD-Ericsson for implementing.
Great to get this over the line!

yxxhero commented:
Awesome

reneleonhardt commented:
Can the tests be improved to cover the unexpected problems and retry this PR?

kickthemooon commented:
I updated my helm version to 3.16.1, tried out `helm dep update` but didn't see an improvement unfortunately. is there something I'm missing? is the lock file required for the improvement?

kickthemooon commented:
Oh so it was reverted, so I'll try 3.16.0

kickthemooon commented:
or could it be that it hasn't been released in any of the versions?

MichaelMorrisEst commented:
It was in 3.16.0 but reverted in 3.16.1 (see https://github.com/helm/helm/pull/13327 for details)

reneleonhardt commented:
Is a fix / new version of these performance improvements planned with better test coverage?

MichaelMorrisEst commented:
> Is a fix / new version of these performance improvements planned with better test coverage?

A new PR has been created to re-introduce the change and fix the issue found in the reverted change. This is waiting for review. See: https://github.com/helm/helm/pull/13342
